packages/utils/src/normalize.ts

```diff
@@ -51,170 +53,173 @@ export function normalizeToSize<T>(
   for (const innerKey in source) {
     // Avoid iterating over fields in the prototype if they've somehow been exposed to enumeration.
     if (!Object.prototype.hasOwnProperty.call(source, innerKey)) {
       continue;
     }
 
-    if (propertyCount >= maxProperties) {
+    if (numAdded >= maxProperties) {
       acc[innerKey] = '[MaxProperties ~]';
       break;
     }
 
-    propertyCount += 1;
+    // Recursively visit all the child nodes
+    const innerValue: unknown = source[innerKey];
+    acc[innerKey] = visit(innerKey, innerValue, depth - 1, maxProperties, memo);
 
-    // Recursively walk through all the child nodes
-    const innerValue: UnknownMaybeWithToJson = source[innerKey];
-    acc[innerKey] = walk(innerKey, innerValue, depth - 1, maxProperties, memo);
+    numAdded += 1;
   }
 
-  // Once walked through all the branches, remove the parent from memo storage
+  // Once we've visited all the branches, remove the parent from memo storage
   unmemoize(value);
 
   // Return accumulated values
   return acc;
 }
 
-/**
- * Transform any non-primitive, BigInt, or Symbol-type value into a string. Acts as a no-op on strings, numbers,
- * booleans, null, and undefined.
- *
- * @param value The value to stringify
- * @returns For non-primitive, BigInt, and Symbol-type values, a string denoting the value's type, type and value, or
- *  type and `description` property, respectively. For non-BigInt, non-Symbol primitives, returns the original value,
- *  unchanged.
- */
-function serializeValue(value: any): any {
-  // Node.js REPL notation
-  if (typeof value === 'string') {
-    return value;
-  }
-
-  const type = Object.prototype.toString.call(value);
-  if (type === '[object Object]') {
-    return '[Object]';
-  }
-  if (type === '[object Array]') {
-    return '[Array]';
-  }
-
-  // `makeSerializable` provides a string representation of certain non-serializable values. For all others, it's a
-  // pass-through.
-  const serializable = makeSerializable(value);
-  return isPrimitive(serializable) ? serializable : type;
-}
+// TODO remove this in v7 (we don't use it anywhere, but it's a public method)
+export { visit as walk };
 
 /**
- * makeSerializable()
+ * Stringify the given value. Handles various known special values and types.
  *
- * Takes unserializable input and make it serializer-friendly.
+ * Not meant to be used on simple primitives which already have a string representation, as it will, for example, turn
+ * the number 1231 into "[Object Number]", nor on `null`, as it will throw.
  *
- * Handles globals, functions, `undefined`, `NaN`, and other non-serializable values.
+ * @param value The value to stringify
+ * @returns A stringified representation of the given value
  */
-function makeSerializable<T>(value: T, key?: any): T | string {
-  if (key === 'domain' && value && typeof value === 'object' && (value as unknown as { _events: any })._events) {
-    return '[Domain]';
-  }
+function stringifyValue(
```

Comment:
This might be a bit opinionated but I would prefer a function `serializeValue(value: unknown, key: unknown): string | number | boolean | null` here that also takes in serializable primitives.

I like that we unified `makeSerializable` and `serializeValue`. I would find that approach just a tiny bit cleaner. This should by the way not block. Just wanted to open up the discussion.

Reply:
> This might be a bit opinionated

Opinions welcome! Can't promise to always agree with them, but def open to hearing what they are!

> I would prefer a function `serializeValue(value: unknown, key: unknown): string | number | boolean | null` here that also takes in serializable primitives.

I actually did have it that way at first. Though neither of these is a huge deal, there are a two reasons I switched to handling primitives before the `stringifyValue` call rather than inside of it (which would then have meant a name change, as you mention):

- If the function can take and return any primitive, I then have to check as soon as I call it what I've gotten back so I can know what to return immediately, and the `if (value === null || (['number', 'boolean', 'string'].includes(typeof value) && !isNaN(value)))` check that's there now instead (before the function call rather than after) is a little less complicated. (Alternatively, if I don't check that immediately after, I lose the simplicity of knowing from that point forward that I've definitely got a string and nothing else.)

- It made naming easier. If it returns primitives, _is_ it really a serializer? Or is it just, as the old function said, ensuring that everything is capable of being serialized later? But then, what of the things which actually _do_ get serialized by it? (This begins to explain how we came to have two separate functions in the first place.) By corralling all of the actually-serialized cases into their own function, I was able call it something which accurately reflects what it's doing.

Reply:
Yeah those points are actually valid. See me convinced.